File: 17-letter-combinations-of-a-phone-number.java
class Solution {
    private static Map<Character, String> phoneLetters = Map.of('2', "abc", 
                                                                '3', "def", 
                                                                '4', "ghi", 
                                                                '5', "jkl", 
                                                                 '6', "mno", 
                                                                '7', "pqrs", 
                                                                '8', "tuv", 
                                                                '9', "wxyz");
    
    public List<String> letterCombinations(String digits) {
        List<String> result = new ArrayList<>();
        if(digits.length() == 0) {
            return result;
        }
        
        backtrack(digits, new StringBuilder(), result);
        return result;
    }
    
    private void backtrack(String digits, StringBuilder path, List<String> result) {
        if(path.length() == digits.length()) {
            result.add(path.toString());
            return;
        }
        
        char nextDigit = digits.charAt(path.length());
        for(char letter : phoneLetters.get(nextDigit).toCharArray()) {
            path.append(letter);
            backtrack(digits, path, result);
            path.deleteCharAt(path.length() - 1);
        }
    }
}

File: 17-letter-combinations-of-a-phone-number.py
Each digit maps to letters. I use backtracking: starting from the first digit, 
I try every mapped letter, append it to the current string, recurse to the next digit, 
and when I reach the end, I add the built string to the result. 
If the input is empty, I return an empty list.

from typing import List

class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []

        digit_to_letters = {
            "2": "abc", "3": "def", "4": "ghi", "5": "jkl",
            "6": "mno", "7": "pqrs", "8": "tuv", "9": "wxyz",
        }

        result = []
        path = []

        def backtrack(i: int) -> None:
            if i == len(digits):
                result.append("".join(path))
                return

            letters = digit_to_letters[digits[i]]
            for ch in letters:
                path.append(ch)
                backtrack(i + 1)
                path.pop()

        backtrack(0)
        return result
